### PR TITLE
[query] Make shallow PCArray/PCBinary copies no cost

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
Would previously pull out address and load length into local variables even if we just need to pass along pointer.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 1]
```diff
@@ -411,62 +411,56 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
...
         Code(
+          srcAddrVar := srcAddress,
+          len := srcArray.loadLength(srcAddrVar),
           newAddr := allocate(region, len),
-          Region.copyFrom(srcAddress, newAddr, contentsByteSize(len)),
+          Region.copyFrom(srcAddrVar, newAddr, contentsByteSize(len)),
           deepPointerCopy(mb, region, newAddr),
           newAddr)
       } else
         srcAddress
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: Still prefer early return :)

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBinary.scala [hunk 2]
```diff
@@ -86,52 +79,54 @@ class PCanonicalBinary(val required: Boolean) extends PBinary {
...
+      val srcAddrVar = mb.newLocal[Long]
+      val len = mb.newLocal[Int]
       val newAddr = mb.newLocal[Long]
       Code(
+        srcAddrVar := srcAddress,
+        len := srcBinary.loadLength(srcAddrVar),
         newAddr := allocate(region, len),
         storeLength(newAddr, len),
-        Region.copyFrom(srcAddress + srcBinary.lengthHeaderBytes, newAddr + lengthHeaderBytes, len.toL),
+        Region.copyFrom(srcAddrVar + srcBinary.lengthHeaderBytes, newAddr + lengthHeaderBytes, len.toL),
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBinary.scala]: Why not copy all bytes. I believe you brought this up yesterday, confirming: Lena can be different between source and dest I think. Could you provide an example?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBinary.scala]: this is the generic interface, I think. There's no guarantee that the length and data are stored together.

**Pull request merged**